File: python/page_rank.py
from pyvis.network import Network
from random import choice

def page_rank(adj_matrix: list[list[int]], iter: int = 1, damp_factor: float = 0.85) -> list[float]:
    """
    Calculates the rank of each page on the given adjacency matrix.

    Parameters:
        matrix (list[list[int]]): Adjacency matrix.
        iter (int):               Algorithm iteration count. Default = 1
        d (float):                Damping factor. Default = 0.85.

    Returns:
        list[float]:              Result list of page ranks.
    """

    # Node count
    n = len(adj_matrix)

    # Initialize pagerank as an uniform vector
    pagerank = [1 / n] * n

    # PageRank algorithm
    for _ in range(iter):
        new_pagerank = [(1 - damp_factor) / n] * n

        for i in range(n):
            for j in range(n):
                if j == i:
                    continue

                new_pagerank[i] += (
                    damp_factor * normal_value(adj_matrix[j][i], sum(adj_matrix[j])) * pagerank[j]
                )

        pagerank = new_pagerank

    return pagerank

# Divisions by zero return 0 in this algorithm
def normal_value(m: int, s: int) -> float:
    return 0 if s == 0 else m / s

# Generates a random matrix
def generate_matrix(n: int) -> list[list[int]]:
    return [[choice([0, 0, 1]) for _ in range(n)] for _ in range(n)]


if __name__ == "__main__":
    net = Network(directed=True)

    # Example matrix
    # matrix = [
    #     [0, 1, 0],
    #     [1, 0, 1],
    #     [1, 0, 0],
    #]

    matrix = generate_matrix(5)

    n = len(matrix)
    pagerank = page_rank(matrix)

    net.add_nodes(
        [i for i in range(n)],
        value=pagerank,
        label=[chr(i + 65) for i in range(n)],
    )

    for i in range(n):
        for j in range(n):
            if matrix[i][j] == 1:
                net.add_edge(i, j, arrows="to")

    net.show("graph.html", notebook=False)

File: golang/internal/pagerank.go
package pagerank

import (
	"errors"
)

/*
Calculates the rank of each page on the given adjacency matrix of type T.

T is any real number.

Parameters:

  adj_matrix [][]T:   Adjacency matrix, always a square matrix.
  iter         int:   Algorithm iteration count. Default = 1
  damp_factor    T:   Damping factor. Default = 0.85.

Returns:
	[]T:                Result list of page ranks.
	error:              Returns in case the matrix lenght is 0
*/

func RankPages[T Real](adj_matrix [][]T, iter int, damp_factor T) ([]T, error) {
	// Node count
	n := len(adj_matrix)

	if n == 0 {
		return nil, errors.New("Matrix has size 0")
	}

	// Normalizes the value.
	// Returns 0 when dividing by zero
	average_or_zero := Divide[T]

	t_n := T(n)

	// Uniform vector inicialization
	pagerank := generate_uniform_vector(n, 1/t_n)

	// PageRank algorithm
	for range iter {
		new_pagerank := generate_uniform_vector(n, (1-damp_factor)/t_n)

		for i := range n {
			for j := range n {
				if i == j {
					continue
				}

				new_pagerank[i] += adj_matrix[j][i] *
					average_or_zero(pagerank[j], Sum(adj_matrix[j]...)) *
					damp_factor
			}
		}

		pagerank = new_pagerank
	}

	return pagerank, nil
}

// Returns an uniform vector of size n, and applies the same result of type T to all elements.
func generate_uniform_vector[T any](n int, v T) []T {
	result_matrix := make([]T, n)

	for i := range result_matrix {
		result_matrix[i] = v
	}

	return result_matrix
}
